src/test/java/org/broadinstitute/hellbender/engine/GATKPathSpecifierUnitTest.java

```diff
@@ -286,6 +287,173 @@ public void testStdOut() throws IOException {
+                {"gs://hellbender/test/resources/aFile.fasta", "aFile"},
+                {"http://bucket/aFile.bam?query=param", "aFile"},
+
+                // This case satisfies hasExtension(".fasta.gz"), but getExtension() returns ".gz".
+                {"aFile.fasta.gz", "aFile.fasta"},
+                // basename is ".fasta"!
+                {".fasta.gz", ".fasta",},
+        };
+    }
+
+    @Test(dataProvider = "getBaseNameTestCases")
+    public void testGetBaseName(final String spec, final String baseName) {
+        Assert.assertEquals(new GATKPathSpecifier(spec).getBaseName(), baseName);
+    }
+
+    @DataProvider(name="negativeGetBaseNameTestCases")
+    public Object[][] negativeGetBaseNameTestCases() {
+        return new Object[][]{
+                // no extensions
+                {"/"},
+                {"."},
+                {"/."},
+                {"/name/.fasta"},
+                {"localFile"},
+                {"gs://hellbender/test/resources"},
+                {"gs://hellbender/test/resources?query=param"},
+                {"gs://hellbender/test/resources/"},
+                {"gs://hellbender/test/resources/?query=param"},
+        };
+    }
+
+    @Test(dataProvider = "negativeGetBaseNameTestCases", expectedExceptions = {IllegalArgumentException.class})
+    public void testNegativeGetBaseName(final String spec) {
+        new GATKPathSpecifier(spec).getBaseName();
+    }
+
+    @DataProvider(name="isFastaTestCases")
+    public Object[][] isFastaTestCases() {
+        final String twoBitRefURL = publicTestDir + "large/human_g1k_v37.20.21.2bit";
+        return new Object[][] {
+                { twoBitRefURL, false },
+                { "file://" + twoBitRefURL, false },
+                { hg38Reference, true }, // gzipped
+                { "file://" + hg38Reference, true }, // gzipped
+                { GCS_b37_CHR20_21_REFERENCE_2BIT, false },
+                { GCS_b37_CHR20_21_REFERENCE, true },
+                // dummy query params at the end to make sure URI.getPath does the right thing
+                { GCS_b37_CHR20_21_REFERENCE + "?query=param", true}
+        };
+    }
+
+    @Test(dataProvider = "isFastaTestCases")
+    public void testIsFasta(final String referenceSpec, final boolean expectedIsFasta) {
+        Assert.assertEquals(new GATKPathSpecifier(referenceSpec).isFasta(), expectedIsFasta);
+    }
+
+    @DataProvider(name="isHadoopURLTestCases")
+    public Object[][] isHadoopURLTestCases() {
+        return new Object[][] {
+                { GATKPathSpecifier.HDFS_SCHEME + "://someFile", true },
+                { "file://someFile.bam", false },
+                { "someFile.bam", false },
+        };
+    }
+
+    @Test(dataProvider = "isHadoopURLTestCases")
+    public void testIsHadoopURL(final String referenceSpec, final boolean expectedIsHadoop) {
+        Assert.assertEquals(new GATKPathSpecifier(referenceSpec).isHadoopURL(), expectedIsHadoop);
+    }
+
```

Comment:
Should we simple tests for isSam, isBam, is Cram ?

Reply:
We've come this far so, yes.